components/zip321-parse/src/lib.rs

```diff
@@ -0,0 +1,903 @@
+use core::fmt::Debug;
+use std::collections::BTreeMap;
+
+use base64::{prelude::BASE64_URL_SAFE_NO_PAD, Engine};
+use nom::{
+    bytes::complete::{tag, take_till},
+    character::complete::{alpha1, char, digit0, digit1, one_of},
+    combinator::{all_consuming, map_opt, map_res, opt, recognize},
+    multi::separated_list0,
+    sequence::{preceded, separated_pair, tuple},
+    AsChar, IResult, InputTakeAtPosition,
+};
+use percent_encoding::percent_decode;
+use snafu::prelude::*;
+
+/// Parsing errors.
+#[derive(Debug, Clone, PartialEq, Eq, Snafu)]
+pub enum Error {
+    /// The ZIP 321 request included more payments than can be created within a single Zcash
+    /// transaction. The wrapped value is the number of payments in the request.
+    #[snafu(display("Cannot create a Zcash transation container {count} payments"))]
+    TooManyPayments { count: usize },
+
+    /// A memo field in the ZIP 321 URI was not properly base-64 encoded
+    #[snafu(display("Memo value was not correctly base64-encoded: {source:?}"))]
+    InvalidBase64 { source: base64::DecodeError },
+
+    /// A memo byte array was too long.
+    #[snafu(display("Memo length {count} is larger than maximum of 512"))]
+    MemoBytesTooLong { count: usize },
+
+    #[snafu(display("Error parsing lead address: {msg}"))]
+    LeadAddress { msg: String },
+
+    #[snafu(display("Error parsing query parameters: {msg}"))]
+    QueryParams { msg: String },
+
+    /// Parsing encountered a duplicate ZIP 321 URI parameter for the returned payment index.
+    #[snafu(display("There is a duplicate {} parameter at index {idx}", param.name()))]
+    DuplicateParameter { param: Box<Param>, idx: usize },
+
+    /// The payment at the wrapped index did not include a recipient address.
+    #[snafu(display("Payment {idx} is missing its recipient address"))]
+    RecipientMissing { idx: usize },
+}
+
+type Result<T, E = Error> = core::result::Result<T, E>;
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+#[repr(transparent)]
+pub struct Memo([u8; 512]);
```

Comment:
Can we not use `zcash_protocol::memo::MemoBytes` directly instead? I know it isn't `repr(transparent)`, but it has `as_array` which should be enough, shouldn't it?

Reply:
`zip321-parse` is being created explicitly to drop the `zcash_protocol` dependency.